radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp

```diff
@@ -0,0 +1,191 @@
+      if (FLYSKY_HALL_PROTOLO_HEAD == ch) {
+        hallBuffer->head = FLYSKY_HALL_PROTOLO_HEAD;
+        hallBuffer->status = GET_ID;
+        hallBuffer->msg_OK = 0;
+      }
+      break;
+
+    case GET_ID:
+      hallBuffer->hallID.ID = ch;
+      hallBuffer->status = GET_LENGTH;
+      break;
+
+    case GET_LENGTH:
+      hallBuffer->length = ch;
+      hallBuffer->dataIndex = 0;
+      hallBuffer->status = GET_DATA;
+      if (0 == hallBuffer->length) {
+        hallBuffer->status = GET_CHECKSUM;
+        hallBuffer->checkSum = 0;
+      }
+      break;
+
+    case GET_DATA:
+      hallBuffer->data[hallBuffer->dataIndex++] = ch;
+      if (hallBuffer->dataIndex >= hallBuffer->length) {
+        hallBuffer->checkSum = 0;
+        hallBuffer->dataIndex = 0;
+        hallBuffer->status = GET_STATE;
+      }
+      break;
+
+    case GET_STATE:
+      hallBuffer->checkSum = 0;
+      hallBuffer->dataIndex = 0;
+      hallBuffer->status = GET_CHECKSUM;
+      // fall through!
+
+    case GET_CHECKSUM:
+      hallBuffer->checkSum |= ch << ((hallBuffer->dataIndex++) * 8);
+      if (hallBuffer->dataIndex >= 2) {
+        hallBuffer->dataIndex = 0;
+        hallBuffer->status = CHECKSUM;
+        // fall through!
+      } else {
+        break;
+      }
+
+    case CHECKSUM:
+      if (hallBuffer->checkSum ==
+          //crc16(CRC_1021, &hallBuffer->head, hallBuffer->length + 3, 0xffff)) {
+            crc8_hw(&hallBuffer->head, hallBuffer->length + 3)) {
+        hallBuffer->msg_OK = 1;
+      }
+      hallBuffer->status = GET_START;
+      break;
+  }
+}
+
+volatile bool _fs_gimbal_detected;
+
+void flysky_gimbal_loop()
+{
+  uint8_t byte;
+
+  while (_fs_get_byte(&byte)) {
+    HallProtocol.index++;
+
+    _fs_parse(&HallProtocol, byte);
+    if (HallProtocol.msg_OK) {
+      HallProtocol.msg_OK = 0;
+      HallProtocol.stickState = HallProtocol.data[HallProtocol.length - 1];
+
+      switch (HallProtocol.hallID.hall_Id.receiverID) {
+        case TRANSFER_DIR_TXMCU:
+          if (HallProtocol.hallID.hall_Id.packetID ==
+              FLYSKY_HALL_RESP_TYPE_VALUES) {
+            int16_t* p_values = (int16_t*)HallProtocol.data;
+            uint16_t* adcValues = getAnalogValues();
+            for (uint8_t i = 0; i < 4; i++) {
+              adcValues[i] = FLYSKY_OFFSET_VALUE - p_values[i];
+            }
+          }
+          break;
+      }
+      _fs_gimbal_detected = true;
+    }
+  }
+}
+
+static void flysky_gimbal_deinit()
+{
+  //STM32SerialDriver.deinit(_fs_usart_ctx);
+}
+
+bool flysky_gimbal_init()
+{
+  /*
+  * This is already initialized in aux_serial_driver.cpp I think
+  etx_serial_init cfg = {
+    .baudrate = FLYSKY_HALL_BAUDRATE,
+    .encoding = ETX_Encoding_8N1,
+    .direction = ETX_Dir_RX,
+    .polarity = ETX_Pol_Normal,
+  };
+  */
+
+  _fs_gimbal_detected = false;
+  //_fs_usart_ctx = auxSerialRxFifo.init(REF_STM32_SERIAL_PORT(FSGimbal), &cfg);
+  //auxSerialRxFifo.setIdleCb(_fs_usart_ctx, flysky_gimbal_loop);
```

Comment:
I was not analyzing original code, but it seems like you are lacking this callback for it to work. `stm32_serial_set_idle_cb` implementation in aux serial driver is currently misssing.

Reply:
Seems like it is already there but I have disabled it due to no purpose for current needs.
Check: `AUX_SERIAL_USART_IRQHandler` and disabled receive part, also see `108: USART_ITConfig(AUX_SERIAL_USART, USART_IT_RXNE, ENABLE);` should be enabled for your needs.

Adding a callback registration fuction and call in receive section should do the trick.

Reply:
Hi Janek, thanks for your comments. Something I just realize is that the first version of my PR was based on the current master of edgeTX which they have moved from STStdPeriphLibs to STM32 LL - Something this repo obviously lacks.

Then I went back to the edgeTX version 2.8.5 and I can see more similarities due to the heavier use of he STStdPeriphLibs. In this version they run a main hall gimbal loop in a 1ms interrupt function.

I think I can easily modify the 5ms interrupt function to run in 1ms and have some prescalers to also run whatever else needs to run every 5ms.

While analyzing all these I was wondering if we could also move I6X repo to utilize the STM32 LL binaries (which are available from STM) and even move the whole code to latest version of edgeTX? Setting aside the obvious mammoth of a task that this would entail, what are your toughts on this matter?

Reply:
As discussed on discord - we will follow the callback solution as it requires much less CPU time.